Sources/AWSSDKSwiftCore/Message/AWSResponse.swift

```diff
@@ -78,4 +80,201 @@ public struct AWSResponse {
+                    outputNode.addChild(node)
+                }
+            }
+            // add status code to output dictionary
+            if let statusCodeParam = Output.statusCodeParam {
+                let node = XML.Element(name: statusCodeParam, stringValue: "\(self.status.code)")
+                outputNode.addChild(node)
+            }
+            return try XMLDecoder().decode(Output.self, from: outputNode)
+
+        case .raw(let payload):
+            if let payloadKey = payloadKey {
+                outputDict[payloadKey] = payload
+            }
+
+        default:
+            break
+        }
+
+        // add header values to output dictionary, so they can be decoded into the response object
+        for (key, value) in self.headers {
+            let headerParams = Output.headerParams
+            if let index = headerParams.firstIndex(where: { $0.key.lowercased() == key.lowercased() }) {
+                // check we can convert to a String. If not just put value straight into output dictionary
+                guard let stringValue = value as? String else {
+                    outputDict[headerParams[index].key] = value
+                    continue
+                }
+                if let number = Double(stringValue) {
+                    outputDict[headerParams[index].key] = number.truncatingRemainder(dividingBy: 1) == 0 ? Int(number) : number
+                } else if let boolean = Bool(stringValue) {
+                    outputDict[headerParams[index].key] = boolean
+                } else {
+                    outputDict[headerParams[index].key] = stringValue
+                }
+            }
+        }
+        // add status code to output dictionary
+        if let statusCodeParam = Output.statusCodeParam {
+            outputDict[statusCodeParam] = self.status.code
+        }
+
+        return try decoder.decode(Output.self, from: outputDict)
+    }
+    
+    /// extract error code and message from AWSResponse
+    func generateError(serviceConfig: ServiceConfig) -> Error? {
+        var errorMessage: ErrorMessage? = nil
+        struct XMLQueryError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
+
+            private enum CodingKeys: String, CodingKey {
+                case code = "Code"
+                case message = "Message"
+            }
+        }
+        struct JSONError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
+
+            private enum CodingKeys: String, CodingKey {
+                case code = "__type"
+                case message = "message"
+            }
+        }
+        struct RESTJSONError: Codable, ErrorMessage {
+            var code: String?
+            var message: String
+
+            private enum CodingKeys: String, CodingKey {
+                case code = "code"
+                case message = "message"
+            }
+        }
```

Comment:
Do we maybe want to move this outside of the function? Would make the whole function easier to read?